File: components/automate-backend-elasticsidecar/habitat/bin/elastic_sidecar.rb
#!/usr/bin/env ruby

require 'bcrypt'
require 'fileutils'
require 'http'
require 'json'
require 'logger'
require 'toml-rb'
require 'mixlib/shellout'
require 'chefstyle'

module AutomateCluster
  class ElasticSidecar
    def initialize
      logger.level = Logger::WARN
      logger.warn 'AutomateCluster::ElasticSidecar initialized!'
    end

    def logger
      @logger ||= Logger.new(STDERR)
    end
    
    def config
      @config ||= load_config
    end
#
    def valid_json?(json)
      JSON.parse(json)
      true
    rescue JSON::ParserError
      false
    end
#
    def test_authentication(user, pass)
      http = HTTP.accept(:json)
                 .basic_auth(user: user, pass: pass)
      http.get('https://localhost:9200', ssl_context: config['ctx_basic'])
    rescue HTTP::Error => e
      logger.error e.message
      'error'
    end
#
    def run_command(cmd)
      environment = { HAB_LICENSE: 'accept-no-persist',
                      JAVA_HOME: config['java_home'] }
      so = Mixlib::ShellOut.new(cmd, environment: environment)
      logger.warn "Running CMD: #{cmd}"
      so.run_command
      so.error!
      so
    rescue Mixlib::ShellOut::ShellCommandFailed => e
      logger.error "encountered exception #{e}!"
      if so.stderr
        message = "\n***********\n"
        message += so.stderr
        message += "***********\n"
      end
      raise message
    end
#
    def load_config
      config_file = '/hab/svc/automate-ha-elasticsidecar/config/elastic_sidecar.toml'
      logger.warn "Attempting to load config from #{config_file}"
      conf = TomlRB.load_file(config_file)
      conf['ssl_mode'] = if conf['ssl_verify'] == false
                           0
                         else
                           1
                         end
#
      conf['ctx_basic'] = OpenSSL::SSL::SSLContext.new
      conf['ctx_basic'].set_params(
        verify_mode: conf['ssl_mode'],
        verify_hostname: conf['ssl_mode'].zero? ? false : true
      )
      conf
    end
#
    def hash_password(password)
      password_hash = BCrypt::Password.create(password)
      password_hash.to_s
    end
#
    def write_internal_users
      puts config.inspect
      internal_users = {
        "_meta" => {
          "type" => "internalusers",
          "config_version" => "2",
        },
        config['admin_username'] => {
          'hash' => hash_password(config['admin_password']),
          'reserved' => true,
          'backend_roles' => [config['admin_username']],
        }
      }
      f = File.new("#{config['securityconfig_path']}/internal_users.yml", 'w+')
      f.write(internal_users.to_yaml)
      f.close
    end
#
    def write_roles_mapping
      roles_mapping = {
        "_meta" => {
          "type" => "rolesmapping",
          "config_version" => "2",
        },
        'all_access' => {
          'reserved' => 'false',
          'backend_roles' => [config['admin_username']],
        },
        'own_index' => {
          'reserved' => 'false',
          'backend_roles' => ['*'],
        }
      }
      f = File.new("#{config['securityconfig_path']}/roles_mapping.yml", 'w+')
      f.write(roles_mapping.to_yaml)
      f.close
    end
#
    def insert_credentials
      insert_command = "#{config['tool_path']}/securityadmin.sh -h #{config['opensearch_ip']} \
        -p #{config['opensearch_port']} -cacert #{config['opensearch_ca']} \
        -cert #{config['admin_cert']} -key #{config['admin_key']} -nhnv -icl \
        -cd #{config['securityconfig_path']}"
      result = run_command(insert_command)
      if result.exitstatus.zero? && result.stdout !~ /ERR: /
        logger.warn 'Credentials successfully rotated'
        logger.warn result
      else
        logger.error "Credentials failed to rotate with error code #{result.exitstatus}"
        logger.error result
      end
    end
#
    def rotate_credentials
      logger.warn 'Preparing to rotate credentials'
      setup_templates
      write_internal_users
      write_roles_mapping
      insert_credentials
    end
#
    def setup_templates
      unless Dir.exist?(config['securityconfig_path'])
        FileUtils.mkdir_p(config['securityconfig_path'])
      end
      FileUtils.cp %w[
        /hab/svc/automate-ha-elasticsidecar/config/securityconfig/action_groups.yml
        /hab/svc/automate-ha-elasticsidecar/config/securityconfig/config.yml
        /hab/svc/automate-ha-elasticsidecar/config/securityconfig/roles.yml
        /hab/svc/automate-ha-elasticsidecar/config/securityconfig/tenants.yml
        /hab/svc/automate-ha-elasticsidecar/config/securityconfig/nodes_dn.yml
        /hab/svc/automate-ha-elasticsidecar/config/securityconfig/whitelist.yml
      ], config['securityconfig_path']
    end
#
    def wait
      logger.debug "Sleeping 60s"
      sleep 60
    end
#
    def run
      loop do
        response = test_authentication(config['admin_username'], config['admin_password'])
        if response.eql? 'error'
          wait
          next
        elsif response.code == 200
          logger.debug "auth successful for #{config['admin_username']}"
        end
#
        case response.code
        when 200
          logger.debug 'Authentication successful, doing nothing'
        when 401, 403
          logger.warn 'Authentication failed, inserting credentials'
          rotate_credentials
        when 503
          if response.body.to_s.eql? 'Open Distro not initialized'
            logger.warn 'Open Distro appears to not be setup, inserting credentials'
            rotate_credentials
          elsif response.body.to_s.valid_json?
            logger.warn 'Auth successful, but elasticsearch appears to be broken, doing nothing.'
          else
            logger.warn 'Elasticsearch returned 503 error with unexpected message'
            logger.warn response.body
          end
        else
          logger.error "Elasticsearch returned #{response.code} code with unexpected message"
          logger.error response.body
        end
        wait
      end
    end
    def run
      loop do
        response = test_authentication(config['admin_username'], config['admin_password'])
        if response.eql? 'error'
          wait
          next
        elsif response.code == 200
          logger.debug "auth successful for #{config['admin_username']} now testing #{config['dashboard_username']}"
          response = test_authentication(config['dashboard_username'], config['dashboard_password'])
        end

        case response.code
        when 200
          logger.debug 'Authentication successful, doing nothing'
          insert_dashboards
        when 401, 403
          logger.warn 'Authentication failed, inserting credentials'
          rotate_credentials
        when 503
          if response.body.to_s.eql? 'Open Distro not initialized'
            logger.warn 'Open Distro appears to not be setup, inserting credentials'
            rotate_credentials
          elsif response.body.to_s.valid_json?
            logger.warn 'Auth successful, but elasticsearch appears to be broken, doing nothing.'
          else
            logger.warn 'Elasticsearch returned 503 error with unexpected message'
            logger.warn response.body
          end
        else
          logger.error "Elasticsearch returned #{response.code} code with unexpected message"
          logger.error response.body
        end
        wait
      end
    end
  end
end

svc = AutomateCluster::ElasticSidecar.new
svc.run

File: components/automate-backend-elasticsidecar/cmd/automate-backend-elasticsidecar/commands/elastic_sidecar.go
package commands

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"time"

	"crypto/tls"
	"net/http"

	"golang.org/x/crypto/bcrypt"
	"gopkg.in/ini.v1"
	"gopkg.in/yaml.v2"
)

type ElasticSidecar struct {
	logger *log.Logger
	config *Config
}

type Config struct {
	AdminUsername      string `ini:"admin_username"`
	AdminPassword      string `ini:"admin_password"`
	JavaHome           string `ini:"java_home"`
	ToolPath           string `ini:"tool_path"`
	OpensearchIP       string `ini:"opensearch_ip"`
	OpensearchPort     string `ini:"opensearch_port"`
	OpensearchCA       string `ini:"opensearch_ca"`
	AdminCert          string `ini:"admin_cert"`
	AdminKey           string `ini:"admin_key"`
	SecurityConfigPath string `ini:"securityconfig_path"`
}

func NewElasticSidecar() *ElasticSidecar {
	logger := log.New(os.Stderr, "", log.LstdFlags)
	logger.Println("AutomateCluster::ElasticSidecar initialized!")
	return &ElasticSidecar{
		logger: logger,
		config: loadConfig(),
	}
}

func loadConfig() *Config {
	cfg := new(Config)
	iniFile, err := ini.Load("/hab/svc/automate-ha-elasticsidecar/config/elastic_sidecar.toml")
	if err != nil {
		log.Fatalf("Failed to read config file: %v", err)
	}
	err = iniFile.MapTo(cfg)
	if err != nil {
		log.Fatalf("Failed to map config: %v", err)
	}
	return cfg
}

func (es *ElasticSidecar) validJSON(jsonStr string) bool {
	var js json.RawMessage
	return json.Unmarshal([]byte(jsonStr), &js) == nil
}

func (es *ElasticSidecar) testAuthentication(user, pass string) (int, string) {
	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // nosemgrep
		},
	}
	req, err := http.NewRequest("GET", "https://localhost:9200", nil)
	if err != nil {
		es.logger.Println(err)
		return 0, "error"
	}
	req.SetBasicAuth(user, pass)
	resp, err := client.Do(req)
	if err != nil {
		es.logger.Println(err)
		return 0, "error"
	}
	defer resp.Body.Close()
	body, _ := ioutil.ReadAll(resp.Body)
	return resp.StatusCode, string(body)
}

func (es *ElasticSidecar) runCommand(cmd string) error {
	cmdExec := exec.Command("bash", "-c", cmd)
	cmdExec.Env = append(os.Environ(), fmt.Sprintf("HAB_LICENSE=accept-no-persist"), fmt.Sprintf("JAVA_HOME=%s", es.config.JavaHome))
	es.logger.Println("Running CMD:", cmd)
	output, err := cmdExec.CombinedOutput()
	if err != nil {
		es.logger.Println("encountered exception:", err)
		es.logger.Println(string(output))
		return err
	}
	return nil
}

func (es *ElasticSidecar) hashPassword(password string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	return string(bytes), err
}

func (es *ElasticSidecar) writeInternalUsers() {
	pwd, _ := es.hashPassword(es.config.AdminPassword)
	internalUsers := map[string]interface{}{
		"_meta": map[string]interface{}{
			"type":           "internalusers",
			"config_version": "2",
		},
		es.config.AdminUsername: map[string]interface{}{
			"hash":          pwd,
			"reserved":      true,
			"backend_roles": []string{es.config.AdminUsername},
		},
		"kibanaserver": map[string]interface{}{
			"hash":     pwd,
			"reserved": true,
		},
		"kibanaro": map[string]interface{}{
			"hash":          pwd,
			"reserved":      false,
			"backend_roles": []string{"kibanauser", "readall"},
		},
		"logstash": map[string]interface{}{
			"hash":          pwd,
			"reserved":      false,
			"backend_roles": []string{"logstash"},
		},
		"readall": map[string]interface{}{
			"hash":          pwd,
			"reserved":      false,
			"backend_roles": []string{"readall"},
		},
		"snapshotrestore": map[string]interface{}{
			"hash":          pwd,
			"reserved":      false,
			"backend_roles": []string{"snapshotrestore"},
		},
	}
	data, _ := yaml.Marshal(internalUsers)
	ioutil.WriteFile(fmt.Sprintf("%s/internal_users.yml", es.config.SecurityConfigPath), data, 0644)
}

func (es *ElasticSidecar) writeRolesMapping() {
	rolesMapping := map[string]interface{}{
		"_meta": map[string]interface{}{
			"type":           "rolesmapping",
			"config_version": "2",
		},
		"all_access": map[string]interface{}{
			"reserved":      false,
			"backend_roles": []string{es.config.AdminUsername},
		},
		"own_index": map[string]interface{}{
			"reserved":      false,
			"backend_roles": []string{"*"},
		},
		"logstash": map[string]interface{}{
			"reserved":      false,
			"backend_roles": []string{"logstash"},
		},
		"kibana_user": map[string]interface{}{
			"reserved":      false,
			"backend_roles": []string{"kibanauser"},
		},
		"readall": map[string]interface{}{
			"reserved":      false,
			"backend_roles": []string{"readall"},
		},
		"manage_snapshots": map[string]interface{}{
			"reserved":      false,
			"backend_roles": []string{"snapshotrestore"},
		},
		"kibana_server": map[string]interface{}{
			"reserved":      true,
			"backend_roles": []string{"kibanaserver"},
		},
	}
	data, _ := yaml.Marshal(rolesMapping)
	ioutil.WriteFile(fmt.Sprintf("%s/roles_mapping.yml", es.config.SecurityConfigPath), data, 0644)
}

func (es *ElasticSidecar) insertCredentials() {
	insertCommand := fmt.Sprintf("%s/securityadmin.sh -h %s -p %s -cacert %s -cert %s -key %s -nhnv -icl -cd %s",
		es.config.ToolPath, es.config.OpensearchIP, es.config.OpensearchPort, es.config.OpensearchCA,
		es.config.AdminCert, es.config.AdminKey, es.config.SecurityConfigPath)
	err := es.runCommand(insertCommand)
	if err != nil {
		es.logger.Println("Credentials failed to rotate with error:", err)
	} else {
		es.logger.Println("Credentials successfully rotated")
	}
}

func (es *ElasticSidecar) rotateCredentials() {
	es.logger.Println("Preparing to rotate credentials")
	es.writeInternalUsers()
	es.writeRolesMapping()
	es.insertCredentials()
}

func (es *ElasticSidecar) wait() {
	es.logger.Println("Sleeping 60s")
	time.Sleep(60 * time.Second)
}

func (es *ElasticSidecar) run() {
	for {
		code, response := es.testAuthentication(es.config.AdminUsername, es.config.AdminPassword)
		if response == "error" {
			es.wait()
			continue
		} else if code == 200 {
			es.logger.Println("auth successful for", es.config.AdminUsername)
		}

		switch code {
		case 200:
			es.logger.Println("Authentication successful, doing nothing")
		case 401, 403:
			es.logger.Println("Authentication failed, inserting credentials")
			es.rotateCredentials()
		case 503:
			if response == "OpenSearch security not initialized" {
				es.logger.Println("OpenSearch security appears to not be setup, inserting credentials")
				es.rotateCredentials()
			} else if es.validJSON(response) {
				es.logger.Println("Auth successful, but OpenSearch appears to be broken, doing nothing.")
			} else {
				es.logger.Println("OpenSearch returned 503 error with unexpected message")
				es.logger.Println(response)
			}
		default:
			es.logger.Println("OpenSearch returned", code, "code with unexpected message")
			es.logger.Println(response)
		}
		es.wait()
	}
}

func Execute() {
	svc := NewElasticSidecar()
	svc.run()
}
